src/webob/acceptparse.py

```diff
@@ -255,11 +294,382 @@ class AcceptEncoding(Accept):
+                )
+            if isinstance(value, (tuple, list)):
+                result = []
+                for element in value:
+                    if isinstance(element, (tuple, list)):
+                        element = _item_qvalue_pair_to_header_element(
+                            pair=element
+                        )
+                    result.append(element)
+                header_str = ', '.join(result)
+            else:
+                header_str = str(value)
+        return header_str
+
+
+class AcceptLanguageValidHeader(AcceptLanguage):
+    """
+    Represent a valid ``Accept-Language`` header.
+
+    A valid header is one that conforms to :rfc:`RFC 7231, section 5.3.5
+    <7231#section-5.3.5>`.
+
+    We take the reference from the ``language-range`` syntax rule in :rfc:`RFC
+    7231, section 5.3.5 <7231#section-5.3.5>` to :rfc:`RFC 4647, section 2.1
+    <4647#section-2.1>` to mean that only basic language ranges (and not
+    extended language ranges) are expected in the ``Accept-Language`` header.
+
+    This object should not be modified. To add to the header, we can use the
+    addition operators (``+`` and ``+=``), which return a new object (see the
+    docstring for :meth:`AcceptLanguageValidHeader.__add__`).
+    """
+
+    # RFC 7231 Section 5.3.5 "Accept-Language":
+    # Accept-Language = 1#( language-range [ weight ] )
+    # language-range  =
+    #           <language-range, see [RFC4647], Section 2.1>
+    # RFC 4647 Section 2.1 "Basic Language Range":
+    # language-range   = (1*8ALPHA *("-" 1*8alphanum)) / "*"
+    # alphanum         = ALPHA / DIGIT
+    lang_range_re = r"""
+        \*|
+        (?:
+        [A-Za-z]{1,8}
+        (?:-[A-Za-z0-9]{1,8})*
+        )
+    """
+    lang_range_n_weight_re = _item_n_weight_re(item_re=lang_range_re)
+    lang_range_n_weight_compiled_re = re.compile(
+        lang_range_n_weight_re,
+        re.VERBOSE
+    )
+    accept_language_compiled_re = _list_1_or_more__compiled_re(
+        element_re=lang_range_n_weight_re,
+    )
+
+    def __init__(self, header_value):
+        """
+        Create an :class:`AcceptLanguageValidHeader` instance.
+
+        :param header_value: (``str``) header value.
+        :raises ValueError: if `header_value` is an invalid value for an
+                            ``Accept-Language`` header.
+        """
+        self._header_value = header_value
+        self._parsed = list(self.parse(header_value))
+        self._parsed_nonzero = [
+            (language_range, qvalue) for language_range, qvalue in self.parsed
+            if qvalue
+        ]
+
+    @property
+    def header_value(self):
+        """(``str``) The header value."""
+        return self._header_value
+
+    @property
+    def parsed(self):
+        """
+        (``list``) Parsed form of the header.
+
+        A list of (language range, quality value) tuples.
+        """
+        return self._parsed
+
+    @classmethod
+    def parse(cls, value):
```

Comment:
> When you say "returns something implementing the AcceptLanguage api", do you mean return an instance of one of the three classes, or an AcceptLanguage instance that, say, proxies any attribute access and method calls to an instance it holds of one of the three subclasses?

It would return one of 3 different instances that conform to the same api.

> If using isinstance is a problem, should there be an is_valid attribute? (What should is_valid return for a NoHeader? It's not really valid or invalid — there is no header at all.) What should be the attribute to indicate that the header is a NoHeader, or should we leave the user to check that .header_value is None?

Yes I think `header_value` might be a good way to inspect things but you could also just define some `status` property that is `valid` or `invalid` or `missing` if anyone really needs to know.

> isn't NilAccept very much an implementation detail

Yes, it was an implementation detail - and I hope we can do something similar here.

As far as docstrings go you can just inherit the `AcceptLanguage` docstrings if you want or you can define them separately for each subclass. In general the user is expected to program against the contract defined in the `AcceptLanguage` class - not the specifics of each subclass.